File: guildmark/src/app/models/invoice.ts
// ---------------------------------------------------------------------------
// Invoices
// ---------------------------------------------------------------------------

export type InvoiceType = "disposal" | "loss" | "sale" | "donation";

export interface TaxInvoice {
  id:                          string;
  invoice_number:              string;
  invoice_type:                InvoiceType;
  invoice_date:                string;
  asset_description:           string;
  serial_number:               string | null;
  original_cost:               number | null;
  book_value_at_disposal:      number | null;
  market_value_at_disposal:    number;
  write_off_amount:            number;
  market_anchor_ebay:          number | null;
  pdf_storage_path:            string | null;
  generated_at:                string;
  // Joined
  model_name?:                 string;
}

File: api/bin/lib/models/invoice.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'json_helpers.dart';

part 'invoice.freezed.dart';
part 'invoice.g.dart';

@freezed
abstract class TaxInvoice with _$TaxInvoice {
  const factory TaxInvoice({
    required String id,
    required String invoiceNumber,
    required String invoiceType,
    @DateOnlyConverter() required DateTime invoiceDate,
    required String assetDescription,
    required double marketValueAtDisposal,
    required double writeOffAmount,
    @IsoDateTimeConverter() required DateTime generatedAt,
    String? serialNumber,
    double? originalCost,
    double? bookValueAtDisposal,
    double? marketAnchorEbay,
    String? pdfStoragePath,
    // Joined from asset row
    String? modelName,
  }) = _TaxInvoice;

  const TaxInvoice._();

  factory TaxInvoice.fromJson(Map<String, dynamic> json) =>
      _$TaxInvoiceFromJson(json);

  factory TaxInvoice.fromRow(Map<String, dynamic> row) => TaxInvoice(
    id: row['id'] as String,
    invoiceNumber: row['invoice_number'] as String,
    invoiceType: enumStr(row['invoice_type']),
    invoiceDate: row['invoice_date'] as DateTime,
    assetDescription: row['asset_description'] as String,
    marketValueAtDisposal:
        numToDoubleOrNull(row['market_value_at_disposal']) ?? 0.0,
    writeOffAmount: numToDoubleOrNull(row['write_off_amount']) ?? 0.0,
    generatedAt: row['generated_at'] as DateTime,
    serialNumber: row['serial_number'] as String?,
    originalCost: numToDoubleOrNull(row['original_cost']),
    bookValueAtDisposal: numToDoubleOrNull(row['book_value_at_disposal']),
    marketAnchorEbay: numToDoubleOrNull(row['market_anchor_ebay']),
    pdfStoragePath: row['pdf_storage_path'] as String?,
    modelName: row['model_name'] as String?,
  );
}
